.. raw:: html
    
    <h3>What are SMaHT Data?</h3>

The Somatic Mosaicism across Human Tissues (SMaHT) Network aims to create a reference catalog of somatic mutations and their patterns across a full spectrum of tissue types from 150 non-diseased donors, applying multiple state-of-the-art experimental assays and computational methods. 

The data generated from the SMaHT Network provides a comprehensive and rich public resource for genomic studies that aim to characterize somatic mosaicism and understand its role in human biology and pathology.   

The primary molecular assays of the SMaHT Network include bulk whole genome sequencing (WGS) using both Illumina-based short-read, and PacBio- or ONT-based long-read sequencing platforms, as well as bulk whole transcriptome sequencing on Illumina (total RNA-Seq) and PacBio (Kinnex) platforms.

.. raw:: html
    
    <h3>SMaHT Data Access</h3>

All data/metadata files generated from the SMaHT Network are available for download from the SMaHT Data Portal to the portal users with appropriate access. For self-registered Data Portal users who are not part of the SMaHT Network, the data are available after official data releases.

**Open Access**: The open-access data/metadata files are available for download after a login as a SMaHT Network member as well as a self-registered Data Portal user who is not part of the SMaHT Network.

**Protected Access**: All sequence data (DNA and RNA), inherited germline variant data, and full donor metadata files are protected-access data under dbGaP. Please see our `Protected Data Access page <https://data.smaht.org/docs/access/getting-dbgap-access>`_ for instructions on how to request access for the protected-access SMaHT data under dbGaP.


.. admonition:: Data Release
   :class: tip

   Production Data are only available to members of the SMaHT Consortium at this time. Please check back for upcoming releases.

|

.. raw:: html

    <br />
    <br />
    <div class="table-responsive">
        <table class="data-availability-table table-sm">
            <thead>
                <tr>
                    <th></th>
                    <th><b>COLO829</b></th>
                    <th><b>HapMap Mixture</b></th>
                    <th><b>LB-LA Fibroblasts and Derived iPSCs</b></th>
                    <th><b>Benchmarking Donors</b></th>
                    <th><b>Production Donors</b></th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <td class="row-label"><b>Sequence Data</b><br/> &lpar;FASTQ, BAM/CRAM&rpar;</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Germline Variants</b><br/></td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Somatic Variants</b><br/>&lpar;all variant types, from individual donor or aggregated samples&rpar;</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                    <td class="open">open</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Gene Expression Profile</b><br/></td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                    <td class="open">open</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Epigenetic Profile</b><br/></td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                    <td class="open">open</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Limited Donor Metadata</b><br/> &lpar;i.e., age, sex, hardy scale&rpar;</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Full Donor Metadata</b><br/> &lpar;e.g., smoking status, environmental exposure, prior clinical history&rpar;</td>
                    <td class="not-applicable">N/A</td>
                    <td class="not-applicable">N/A</td>
                    <td class="not-applicable">N/A</td>
                    <td class="protected">protected</td>
                    <td class="protected">protected</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Reference Files</b><br/> &lpar;e.g., Human genome reference, gene models, genome stratification into easy/difficult/extreme-to-map regions&rpar;</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                    <td class="open">open</td>
                </tr>
                <tr>
                    <td class="row-label"><b>Histology</b><br/> &lpar;Amperio SVS images, histology and pathology data&rpar;</td>
                    <td class="not-applicable">N/A</td>
                    <td class="not-applicable">N/A</td>
                    <td class="not-applicable">N/A</td>
                    <td class="not-applicable">N/A</td>
                    <td class="open">open</td>
                </tr>
            </tbody>
        </table>
    </div>