==============================================
Understanding SMaHT Data Portal Manifest Files
==============================================

Please use the ``manifest`` file to download data from the SMaHT Data Portal.

It passes portal access credentials to the command provided in the ``smaht_manifest`` files downloaded from the portal.

-------------
File Manifest
-------------

In the ``File Manifest``, multiple values in a field under a column are separated by the comma (``,``) character.

Below are the columns listed in the ``File Manifest`` files as of the June 2025 data release.

#. **File Download URL** - This URL calls an API that authorizes the user and redirects to a pre-signed URL to download the file.

#. **File Accession** - This value is generated by the SMaHT data portal, it is a unique value  except for extra files associated with a actual file will have the same accession but different file extension. E.g. When a BAM file (*.bam) is selected to download, an index file (*.bai) is the associated extra file that is also downloaded along with it.

#. **File Name** - This value is a file name that also serves as a unique identifier of the file. The file nomenclature schema is described `here <https://data.smaht.org/docs/additional-resources/sample-file-nomenclature>`_.

#. **FileSet Accession** - Unique value that can be used to merge reference rows in the Experimental Manifest.

#. **Analyte Accessions** - Unique value that can be used to merge reference rows in the Analyte Manifest (may contain multiple values).

#. **Sample Accessions** - Unique value that can be used to merge reference rows in the Biosample Manifest (may contain multiple values).

#. **Donor Accession** - Unique value that can be used to merge reference rows in the Donor Manifest.

#. **File Status** - The status of the file (e.g. ``released`` or ``retracted``).

#. **Retracted Reason** - The reason a file was retracted, dependent on File Status.

#. **Size** - File size in bytes.

#. **md5sum** - md5 of the file content.

#. **Data Type** - This value tells you the file type, e.g. ``Aligned Reads``, ``Unaligned Reads`` or ``Variant Calls``.

#. **File Format** - Format of the file (e.g. bam, fastq.gz).

#. **Sample Name** - Sample identifier in SMaHT nomenclature. Please refer to the file nomenclature schema described `here <https://data.smaht.org/docs/additional-resources/sample-file-nomenclature>`_.

#. **Sample Studies** - Studies associated with this file; Benchmarking or Production.

#. **Sample Tissues** - Tissues used to generate this file, if applicable.

#. **Sample Donors** - Similarly, the donors from which the above tissues were generated.

#. **Sample Source** - Sample name provided by a data submitter. If the file is generated from a mixture of samples (e.g. HapMap mix, COLO829-BLT), multiple sample sources will be found here, delimited by ``|``.

#. **Analytes** - Analytes used for analysis, e.g. one of ``RNA``, ``DNA``.

#. **Sequencer** - Name of the sequencing platform used to generate the raw sequencing data e.g. ``PacBio Revio``.

#. **Assay** - Experimental assay used to generate this file, e.g. ``WGS, PCR Free``.

#. **Software Name/Version** - Name and version of software used to generate this file, e.g. ``pbmm2 (1.13.0)``.

#. **Reference Genome** - Reference Genome version used for the analysis, e.g. ``GRCh38 [Official]``.

#. **Final QC Status** - The final QC status of the analysis pipeline used to generate the file (e.g. Pass/Warn or Fail).

#. **QC Comments** - Notable comments on quality control of the file that warrant attention, although they did not result in retraction of the file.

#. **QC Notes** - Values for key metrics that did not pass QC.

#. **File Group** - This field indicates a group of BAM files that can be merged. BAM files with the identical file group value can be merged. Please see the dedicated section below for more information.


File Group
----------

The ``File Group`` field is a special field that indicates which BAM files can be merged. To efficiently process and store large BAMs with high sequencing coverage, the alignment pipeline at DAC produces BAMs per library. To identify BAMs to merge, obtain the files where the file format is BAM and the File Group values are identical.

**NOTE**: ``File Group`` is not generated for Single-Cell and Microbulk files.

Specifically, the ``File Group`` combines several pieces of information, including:

* The center that submitted the raw sequencing data
* Aggregated sample source information
* Aggregated sequencing platform information
* Aggregated experimental assay information

For example:

File Group = ``bcm_gcc-WASHU_CELL-CULTURE-MIXTURE_SMAHT_CORIELL_POOL1-pacbio_revio_hifi-Single-end-17500-no-flow-cell-bulk_wgs_pcr_free``

* ``bcm_gcc`` = Submission center which indicates that ``BCM-GCC`` submitted the sequencing data.
* ``WASHU_CELL-CULTURE-MIXTURE_SMAHT_CORIELL_POOL1`` = Sample Source which indicates this file was generated from SMAHT CORIELL POOL1 sample source, a name designated by the data submitter at BCM.
* ``pacbio_revio_hifi-Single-end-17500-no-flow-cell`` = Sequencing, which indicates that this file was generated from a PacBio Revio sequencer with target read length 17500 and no flow cell information.
* ``bulk_wgs_pcr_free`` = Experimental assay.

The optional ``group tag`` may appear at the end of the ``File Group`` and is used to differentiate files that should not be merged with other files that would otherwise be in the same ``File Group`` based off of the other four values. This may be due to QC issues, or various other reasons.

-------------------------
Additional Manifest Files
-------------------------

Additional files are available to download that contain the full metadata for files. They are separated into four files based on the type of metadata they contain:

* Donor Manifest
* Biosample Manifest
* Analyte Manifest
* Sequencing Manifest

More information on the columns contained in these files can be found in the `Submission Data Dictionary <https://data.smaht.org/docs/submission/submission-data-dictionary>`_ by searching for the item type included as the first part of the column name (e.g. for **AnalyteMolecule**, you could search for ``Analyte`` and look for Molecule under the Metadata Property column)

--------------
Donor Manifest
--------------

The Donor manifest contains metadata on clinical and life history information for individual donors. Currently the metadata for donors is in the Bulk Donor Metadata file, which contains the full metadata for all SMaHT Benchmarking and Production donors. This file can be downloaded from the `Production Browse By File <https://data.smaht.org/browse/?type=File&sample_summary.studies=Production&status=released>`_ page or from individual Donor Overview pages.
This file contains information from the following portal item types:

* ``Donor``
* ``Demographic``
* ``Diagnosis``
* ``DeathCircumstances``
* ``Exposure``
* ``FamilyHistory``
* ``MedicalHistory``
* ``MedicalTreatment``
* ``TissueCollection``


Handling Multiple Values
------------------------

Some columns in the ``Donor Manifest`` contain multiple values.

``Lists``: Columns that contain a list of multiple values associated with a single entry.

Example:

A donor will have one data entry for ``Medical History``.The column **MedicalHistory.allergens** may contain multiple values for multiple allergens. These values will be delimited by the semicolon (“;”) character (e.g. “Pollen;Cats;Penicillin”)

``Nested Lists``: Columns that contain a list of multiple values associated with multiple distinct data entries.

Example:

A donor may have multiple data entries for ``Family History``. The column **FamilyHistory.relatives** can contain multiple values for different relatives that had a history of a particular disease. These values will be delimited by a combination of the pipe (“\|") and the semicolon (“;”) characters. The order of values separated by “\|” indicates how different related columns correspond to one another. For Family History items there are two associated columns:

* **FamilyHistory.disease**: “Diabetes | Breast Cancer”
* **FamilyHistory.relatives**: “Father | Mother;Sister”

In this case, the donor's father had a history of diabetes and the donor's mother and sister had a history of breast cancer. 


Hardy Scale
-----------

The values in the **Donor.hardy_scale** column have the following interpretation:

1 = Violent and fast death. Deaths due to accident, blunt force trauma or suicide, terminal phase estimated at \< 10 min.

2 = Fast death of natural causes. Sudden unexpected deaths of people who had been reasonably healthy, after a terminal phase estimated at \< 1 hr (with sudden death from a myocardial infarction as a model cause of death for this category).

3 =  Intermediate death. Death after a terminal phase of 1 to 24 hrs (not classifiable as 2 or 4); patients who were ill but death was unexpected.

4 = Slow death. Death after a long illness, with a terminal phase longer than 1 day (commonly cancer or chronic pulmonary disease); deaths that are not unexpected.

0 = Ventilator Case. All cases on a ventilator immediately before death.


Exposure Frequency Category
---------------------------

The values in the **Exposure.frequency_category** column have the following interpretation:

For Alcohol Exposure:

* Social = A few drinks per year
* Light = \<1 drink per day (0-6 per week)
* Moderate = Up to 2 drinks per day (7-14 per week)
* Heavy = ≥3 per day (≥15 per week)

For Tobacco Exposure:

* Social = 0-10 cigarettes in past 5yr; ≤6 cigars/pipes per year
* Light = 0-5 cigarettes per day (up to 1/4 ppd); ≤12 cigars/pipes per year
* Moderate = 6-9 cigarettes per day (1/4 to 3/4 ppd); 3-5 cigars/pipes per week
* Heavy = ≥20 cigarettes per day (≥1 ppd); ≥3-5 cigars/pipes per week


------------------
Biosample Manifest
------------------

The ``Biosample Manifest`` contains metadata on the sample(s) a file was derived from. The metadata provided in each column will depend on the **SampleType** and **SampleSourceType**. It contains information from the following portal items:

* ``Sample`` (``TissueSample``, ``CellCultureSample``, and ``CellSample``)
* ``SampleSource`` (``Tissue``, ``CellCulture``, and ``CellCultureMixture``)
* ``CellLine``


----------------
Analyte Manifest
----------------

The ``Analyte Manifest`` contains metadata on the analyte(s) a file was derived from. It contains information from the following portal items:

* ``Analyte``
* ``AnalytePreparation``
* ``PreparationKit``
* ``Treatment``


-------------------
Sequencing Manifest
-------------------

The ``Sequencing Manifest`` contains metadata on the library preparation and sequencing used to generate a file. It contains information from the following portal items:

* ``Library``
* ``LibraryPreparation``
* ``PreparationKit``
* ``Treatment``
* ``Sequencing``


*Please note this functionality is provisional and subject to change. If you encounter issues with this functionality, please report it to DAC!*
